File: cpp/food-chain/food_chain.cpp
//
//
//
#include <vector>
#include <numeric>

#include "food_chain.h"

namespace food_chain {
    const static std::vector<std::string> animals {
        "fly", "spider", "bird", "cat", "dog", "goat", "cow", "horse"
    };

    const static std::vector<std::string> second {
        "",
        "It wriggled and jiggled and tickled inside her.\n",
        "How absurd to swallow a bird!\n",
        "Imagine that, to swallow a cat!\n",
        "What a hog, to swallow a dog!\n",
        "Just opened her throat and swallowed a goat!\n",
        "I don't know how she swallowed a cow!\n",
        "She's dead, of course!\n"
    };

    const static std::string last {
        "I don't know why she swallowed the fly. Perhaps she'll die.\n"
    };

    std::string verse(unsigned long number) {
        auto index = number - 1;

        std::string result = "I know an old lady who swallowed a " + animals.at(index) + ".\n";

        result += second.at(index);

        if (number < 8) {
            for (int i = number; i > 1; i--) {
                result += "She swallowed the " + animals[i - 1] + " to catch the " + animals[i - 2];
                if (i == 3) {
                    result += " that wriggled and jiggled and tickled inside her";
                }

                result += ".\n";
            }

            result += "I don't know why she swallowed the fly. Perhaps she'll die.\n";
        }

        return result;
    }

    std::string verses(unsigned long lower, unsigned long upper) {
        std::string result;

        for (auto i = lower; i <= upper; i++) {
            result += verse(i) + "\n";
        }

        return result;
    }

    std::string sing() {
        return verses(1, 8);
    }
}

File: go/food-chain/food_chain.go
package foodchain

const testVersion = 3

var animals = []string{
	"", "fly", "spider", "bird", "cat", "dog", "goat", "cow", "horse",
}
var first = "I know an old lady who swallowed a "
var second = []string{
	"",
	"",
	"It wriggled and jiggled and tickled inside her.\n",
	"How absurd to swallow a bird!\n",
	"Imagine that, to swallow a cat!\n",
	"What a hog, to swallow a dog!\n",
	"Just opened her throat and swallowed a goat!\n",
	"I don't know how she swallowed a cow!\n",
	"She's dead, of course!",
}
var third = " that wriggled and jiggled and tickled inside her"
var last = "I don't know why she swallowed the fly. Perhaps she'll die."

// Verse returns the index-th sentence of the song
func Verse(index int) string {
	result := first + animals[index] + ".\n" + second[index]

	if index < 8 {
		for i := index; i > 1; i-- {
			result += "She swallowed the " + animals[i] + " to catch the " + animals[i-1]
			if i == 3 {
				result += third
			}
			result += ".\n"
		}

		result += last
	}

	return result
}

// Verses returns lower-th to upper-th sentence of the song
func Verses(lower, upper int) string {
	var result string

	for i := lower; i < upper; i++ {
		result += Verse(i) + "\n\n"
	}

	result += Verse(upper)

	return result
}

// Song returns the whole song
func Song() string {
	return Verses(1, 8)
}
